app/grandchallenge/components/serializers.py

```diff
@@ -26,32 +31,139 @@ class SimpleImageSerializer(serializers.ModelSerializer):
             "value",
             "file",
             "image",
             "pk",
+            "upload_session",
         ]
 
+    def validate(self, attrs):  # noqa: C901
+        try:
+            interface = ComponentInterface.objects.get(
+                slug=attrs["interface_slug"]
+            )
+        except ObjectDoesNotExist:
+            raise serializers.ValidationError(
+                f"Component interface {attrs['interface_slug']} does not exist."
+            )
+
+        attrs.pop("interface_slug")
+        attrs["interface"] = interface
+
+        def get_value():
+            value = attrs.get("value", None)
+            if not value:
+                raise serializers.ValidationError(
+                    f"Value is required for interface kind {interface.kind}"
+                )
+            return value
+
+        def validate_simple():
+            kind_to_type = {
+                ComponentInterface.Kind.INTEGER: int,
+                ComponentInterface.Kind.STRING: str,
+                ComponentInterface.Kind.BOOL: bool,
+                ComponentInterface.Kind.FLOAT: float,
+            }
+            value = get_value()
+            if not isinstance(value, kind_to_type.get(interface.kind)):
+                raise serializers.ValidationError(
+                    f"Type of {value} does not match interface kind {interface.kind}"
+                )
+
+        def validate_annotations():
+            value = get_value()
+            allowed_types = [{"$ref": f"#/definitions/{interface.kind}"}]
+
+            JSONSchemaValidator(
+                schema={**ANSWER_TYPE_SCHEMA, "anyOf": allowed_types}
+            )(value)
+
+        def validate_image():
+            # either image or upload_session should be provided
+            if not any(key in attrs for key in ("upload_session", "image")):
+                raise serializers.ValidationError(
+                    f"Upload_session or image are required for interface kind {interface.kind}"
+                )
+
+        if interface.kind in InterfaceKind.interface_type_simple():
+            validate_simple()
+        if interface.kind in InterfaceKind.interface_type_annotation():
+            validate_annotations()
+        if interface.kind in InterfaceKind.interface_type_image():
+            validate_image()
+
+        return attrs
+
+    def validate_upload_session(self, value):
+        user = self.context.get("user")
+
+        if not user.has_perm("view_rawimageuploadsession", value):
+            raise serializers.ValidationError(
+                f"User does not have permission to use {value}"
+            )
+
+        if value.status is not RawImageUploadSession.PENDING:
+            raise serializers.ValidationError(
+                f"{value} is not ready to be used"
+            )
+        return value
+
+    def validate_image(self, value):
+        user = self.context.get("user")
+
+        if not user.has_perm("view_image", value):
```

Comment:
I wonder if this is too broad, but right now it's probably all we have. Users can get permission to view an image in a variety of ways, and that may not necessarily mean that they can use the image in any way they like. I guess right now they could always download the image and then upload it again as part of an algorithm experiment, so this is adding something new. So, no need to change anything, but I wanted to document it.